substrate-node/pallets/pallet-smart-contract/src/migrations/v8.rs

```diff
@@ -0,0 +1,149 @@
+        debug!("current pallet version: {:?}", PalletVersion::<T>::get());
+        assert!(PalletVersion::<T>::get() >= types::StorageVersion::V8);
+
+        debug!(
+            "👥  Smart Contract pallet to {:?} passes POST migrate checks ✅",
+            PalletVersion::<T>::get()
+        );
+
+        Ok(())
+    }
+}
+
+pub fn migrate_to_version_8<T: Config>() -> frame_support::weights::Weight {
+    debug!(
+        " >>> Starting contract pallet migration, pallet version: {:?}",
+        PalletVersion::<T>::get()
+    );
+
+    let mut reads = 0;
+    let mut writes = 0;
+
+    let mut twin_contract_locked_balances: BTreeMap<u32, BalanceOf<T>> = BTreeMap::new();
+    // Fetch all locked balances based on the contract locks in storage and accumulate them by twin id
+    for (ctr_id, l) in ContractLock::<T>::iter() {
+        let ctr = Contracts::<T>::get(ctr_id);
+        reads += 1;
+        match ctr {
+            Some(contract) => {
+                reads += 1;
+                if !pallet_tfgrid::Twins::<T>::contains_key(contract.twin_id) {
+                    debug!(
+                        "twins: {} does not exist, removing contract and lock",
+                        contract.twin_id
+                    );
+                    Contracts::<T>::remove(ctr_id);
+                    ContractLock::<T>::remove(ctr_id);
+                    writes += 2;
+                } else {
+                    twin_contract_locked_balances
+                        .entry(contract.twin_id)
+                        .and_modify(|v| *v += l.amount_locked)
+                        .or_insert(BalanceOf::<T>::saturated_from(0 as u128));
+                }
+            }
+            None => {
+                debug!(
+                    "no contract found for contract lock {}, cleaning up lock..",
+                    ctr_id
+                );
+                writes + 1;
+                ContractLock::<T>::remove(ctr_id);
+            }
+        }
+    }
+
+    for (twin_id, t) in pallet_tfgrid::Twins::<T>::iter() {
+        reads += 1;
+
+        // If the twin needs to have some locked balance on his account because of running contracts
+        // Check how much we can actually lock based on his current total balance
+        // this will make sure the locked balance will not exceed the total balance on the twin's account
+        let should_lock = match twin_contract_locked_balances.get(&twin_id) {
+            Some(b) => {
+                // get the total balance of the twin - minimum existence requirement
+                debug!("contract locked balance on twin {} account: {:?}", t.id, b);
+                reads += 1;
+                Some(
+                    (<T as Config>::Currency::total_balance(&t.account_id)
+                        - <T as Config>::Currency::minimum_balance())
+                    .min(*b),
+                )
+            }
+            None => None,
+        };
```

Comment:
match here is used to change `Option<T>` to `Option<U>`, so this can be done by `Option::map(...)`

Reply:
![download](https://user-images.githubusercontent.com/17138347/220344393-7625cefd-8f09-4c3c-a78e-226bb0ac3d53.jpeg)